subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -0,0 +1,792 @@
+    GstClockTime prev_dts = GST_CLOCK_TIME_NONE;
+    GstClockTime pts = GST_BUFFER_PTS (nalu->nalu_buf);
+    GstClockTime dts = GST_BUFFER_DTS (nalu->nalu_buf);
+    if (prev_nalu) {
+      prev_pts = GST_BUFFER_PTS (prev_nalu->nalu_buf);
+      prev_dts = GST_BUFFER_DTS (prev_nalu->nalu_buf);
+    }
+    gboolean new_ts = (prev_pts != pts) || (prev_dts != dts);
+
+    nalu->au_start = new_ts /*|| discont */ ;
+    if (prev_nalu && nalu->au_start) {
+      prev_nalu->au_end = TRUE;
+      GST_DEBUG_OBJECT (rtph266pay, "AU start found -> previous AU finished");
+    }
+  }
+
+  // In some cases, we already know where is the end of an AU. Mark it now to
+  // avoid waiting for the next NALU.
+  if (last_nalu) {
+    gboolean au_alignment = rtph266pay->alignment == ALIGNMENT_AU;
+    gboolean marker =
+        GST_BUFFER_FLAG_IS_SET (last_nalu->nalu_buf, GST_BUFFER_FLAG_MARKER);
+    last_nalu->au_end = last_nalu->au_end || au_alignment || marker;
+  }
+}
+
+static GstFlowReturn
+gst_rtp_h266_pay_push_pending_data (GstRtpH266Pay * rtph266pay, gboolean eos)
+{
+  GQueue *nalus = &rtph266pay->nalus;
+  GstFlowReturn ret = GST_FLOW_OK;
+  Nalu *nalu;
+
+  // TODO: handle AP
+  // TODO: send XPS if needed
+
+  // TODO: Maybe it's worth to avoid iterating the NALU list twice, maybe not
+  gst_rtp_h266_pay_set_au_boundaries (rtph266pay);
+
+  // Try to push all NALUs
+  while ((nalu = g_queue_pop_head (nalus))) {
+    gboolean is_last = g_queue_get_length (nalus) == 0;
+
+    // Can't push the last NALU without knowing if it's the end of an AU,
+    // because setting the M bit could be necessary. But have to push it if
+    // we're on EOS
+    if (!eos && is_last && !nalu->au_end) {
+      GST_DEBUG_OBJECT (rtph266pay, "Keeping last NALU: %" NALU_PTR_FORMAT,
+          NALU_ARGS (nalu));
+      g_queue_push_head (nalus, nalu);
+      break;
+    }
+
+    if (gst_rtp_h266_pay_has_to_send_xps (rtph266pay, nalu))
+      gst_rtp_h266_pay_queue_xps_cache (rtph266pay);
+
+    GST_DEBUG_OBJECT (rtph266pay, "Pushing NALU: %" NALU_PTR_FORMAT,
+        NALU_ARGS (nalu));
+
+    if (gst_rtp_h266_pay_up_fits_in_mtu (rtph266pay, nalu))
+      ret = gst_rtp_h266_pay_push_up (rtph266pay, nalu);
+    else
+      ret = gst_rtp_h266_pay_push_fu (rtph266pay, nalu);
+
+    nalu_free (nalu);
+    if (ret != GST_FLOW_OK)
+      break;
+  }
+
+  return ret;
+}
+
+static GstFlowReturn
+gst_rtp_h266_pay_push_up (GstRtpH266Pay * rtph266pay, const Nalu * nalu)
+{
+  GstRTPBasePayload *rtpbasepay = GST_RTP_BASE_PAYLOAD (rtph266pay);
+  GstRTPBuffer rtp = GST_RTP_BUFFER_INIT;
+  GstBuffer *out_buf;
+
+  // Allocate just the RTP header. We'll add the buffer payload later. This way
+  // we avoid unnecessary copies
+  out_buf = gst_rtp_base_payload_allocate_output_buffer (rtpbasepay, 0, 0, 0);
+  if (!gst_rtp_buffer_map (out_buf, GST_MAP_WRITE, &rtp))
+    goto error;
+
+  // Copy buffer metadata
+  gst_buffer_copy_into (out_buf, nalu->nalu_buf,
```

Comment:
Handle error, also gst_buffer_append may return NULL.

Reply:
We're only coping the flags and the timestamps; I think it's not that critical to abort. And because we'll already receive a warning in case of failure, I don't think it's worth it to deal with the returned value here.

> also gst_buffer_append may return NULL.

I'll add a `g_assert()`